Rework "Updates" table checkbox selection controls

janfaracik opened the pull request:
https://user-images.githubusercontent.com/43062514/177877649-53f09952-0736-4e9c-a089-2eb98e194281.mov

This PR adds a reusable table header checkbox component (_working title_ `"row-selection-controller"` - if anybody has a better name for it **please** share) which will select/unselect all checkboxes in a table. It's also extensible and can have custom selection options, e.g. "Compatible plugins" for the Updates page.

For the Updates page, this replaces the controls at the bottom of the page.

### Proposed changelog entries

* Rework "Updates" table checkbox selection controls.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@jenkinsci/sig-ux 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6806"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

janfaracik committed:
Init

janfaracik committed:
Add component to Updates page

janfaracik committed:
Cleanup

janfaracik committed:
Update table.less

janfaracik committed:
Rename and cleanup

janfaracik commented on war/package.json:
```diff
@@ -29,7 +29,7 @@
     "copy-webpack-plugin": "^5.1.1",
     "css-loader": "^3.2.0",
     "css-minimizer-webpack-plugin": "^1.0.0",
-    "eslint": "^7.1.0",
```
I've had to update `eslint` due to my use of optional chaining (e.g. `moreOptionsAllButton?.addEventListener`)

timja commented:
Looks to be a z-index issue:
<img width="377" alt="image" src="https://user-images.githubusercontent.com/21194782/177880524-81e62c12-682a-4b4c-a71a-2ab7ff039675.png">

timja commented:
Selecting compatible is a bit weird atm. It doesn't tick the top checkbox, when you click the top checkbox it does clear selection as expected but you don't get any button feedback

timja commented:
Why does select none need to be a thing? shouldn't that be replaced by clicking the check box again, that's fairly standard behaviour

janfaracik commented:
> Why does select none need to be a thing? shouldn't that be replaced by clicking the check box again, that's fairly standard behaviour
> 
> <img alt="image" width="250" src="https://user-images.githubusercontent.com/21194782/177881367-3f094b30-d421-4f14-bb00-2d03791daf1d.png">

Clicking the checkbox does check/uncheck all. For the design I was essentially inspired by Gmail who do something similar - 
<img width="170" alt="Screenshot 2022-07-08 at 11 28 41" src="https://user-images.githubusercontent.com/43062514/177975192-c3a63eb5-e3c4-4754-96e2-17101649323f.png">

janfaracik committed:
Fix z-index

janfaracik commented:
> Selecting compatible is a bit weird atm. It doesn't tick the top checkbox, when you click the top checkbox it does clear selection as expected but you don't get any button feedback
> 
> From just clicking plugins one by one it does this: <img alt="image" width="83" src="https://user-images.githubusercontent.com/21194782/177882242-d8b4dd03-efcf-4627-b625-fe2557165e26.png">
> 
> which I assume is what compatible should do

Hopefully I'm understanding you correctly - the header checkbox has three states (checked, indeterminate, none), if you click 'Compatible' and not all plugins are checked the header checkbox state will be indeterminate. If all plugins are checked when you click 'Compatible' then the header checkbox state will be checked.

Might have to look into reworking it if the behaviour isn't clear

timja commented:
Yes but compatible didn't seem to be setting the half state of the top checkbox.

clicking them individually did set that correctly, selecting compatible didn't set any state on it which looked wrong

yaroslavafenkin commented:
I observe some weird behaviour when there are no updates available.
I've checked the checkbox and now I cannot change the state:
![image](https://user-images.githubusercontent.com/91559310/178009064-298bc82c-4791-40ea-a694-6cac61ca8142.png)
Selecting any of the items in the dropdown or trying to uncheck does nothing.

janfaracik committed:
Disable control if there are no checkboxes to manage

janfaracik committed:
Fix 'Compatible plugins' not updating the table header checkbox

timja reviewed (commented):
all issues I found are fixed, let's run through ATH to be safe though

NotMyFault commented:
> all issues I found are fixed, let's run through ATH to be safe though

ATH is green, one test failure remains on this PR at `hudson.PluginManagerInstalledGUITest.test_enable_disable_uninstall`.

janfaracik committed:
Merge branch 'master' into update-table-selection

janfaracik committed:
Update yarn.lock

janfaracik committed:
Potential fix for test

timja committed:
Merge branch 'master' into update-table-selection

timja committed:
Fix master build (lint failure)

timja committed:
Merge branch 'fix-lint' into update-table-selection

timja committed:
Fix test

janfaracik committed:
Merge branch 'master' into update-table-selection

timja reviewed (approved):
LGTM, I re-tested locally after all the merges

NotMyFault reviewed (approved):
Looks awesome, apologises for my conflicts caused 🥴

uhafner commented:
🚀 Thanks a lot for fixing this usability problem! 🚀 

What are compatible plugins? Is this just a demo or what are the semantics of compatible?

NotMyFault commented:
> 🚀 Thanks a lot for fixing this usability problem! 🚀
> 
> What are compatible plugins? Is this just a demo or what are the semantics of compatible?

That'd be the case if a plugin uses a higher jenkins.version than the one you have installed, has unmet or faulty third party dependencies afaik.

(Though I'm not too sure if that is still up with the times, the UC doesn't let you update plugins to a jenkins.version you don't use.)

timja commented:
> That'd be the case if a plugin uses a higher jenkins.version than the one you have installed, has unmet or faulty third party dependencies afaik.
> 
> (Though I'm not too sure if that is still up with the times, the UC doesn't let you update plugins to a jenkins.version higher than your current version.)

No that's a different mechanism

This is https://wiki.jenkins-ci.org/display/JENKINS/Marking+a+new+plugin+version+as+incompatible+with+older+versions

`hpi.compatibleSinceVersion` is the plugin pom property.

I tested the functionality by installing the slack plugin version 1.0 and then checking for updates.

uhafner commented:
I see. Since this option makes only sense if there are actually rows that are marked as incompatible I would suggest to not show this option if there are no such rows available.

janfaracik committed:
Add tooltip to Compatible button to explain what it does

janfaracik committed:
Hide checkbox controller dropdown if all plugins are compatible

janfaracik commented:
I've updated the branch to hide the dropdown if all updates available are compatible (thanks @uhafner!) and I've also added a small tooltip the Compatible menu item,` "Selects plugin updates that are compatible with their current version"`. If anything/the label could be improved/changed let me know :)

Thanks

uhafner reviewed (approved):
I haven't looked at the code, but the UI looks good now!

NotMyFault commented:
This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

NotMyFault commented:
> > This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
> > Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
> > Thanks!
> 
> Heya!
> 
> Is there a blocker on merging at the moment? Got a [few PRs](https://github.com/jenkinsci/jenkins/pulls?q=is%3Apr+is%3Aopen+label%3Aready-for-merge+author%3Ajanfaracik) that are ready to merge but haven't been yet.
> 
> Thanks :)

Not that I'm aware of. Speaking for the quoted comment, it's meant to start the timer, I'm not eligible to merge PRs once the grace period expired.

janfaracik commented:
> > > This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
> > > Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
> > > Thanks!
> > 
> > 
> > Heya!
> > Is there a blocker on merging at the moment? Got a [few PRs](https://github.com/jenkinsci/jenkins/pulls?q=is%3Apr+is%3Aopen+label%3Aready-for-merge+author%3Ajanfaracik) that are ready to merge but haven't been yet.
> > Thanks :)
> 
> Not that I'm aware of. Speaking for the quoted comment, it's meant to start the timer, I'm not eligible to merge PRs once the grace period expired.

Ah gotcha! Thanks @NotMyFault and @timja :)

NotMyFault commented:
https://github.com/jenkinsci/jenkins/pull/6806/commits/86ad254472fc4cacfc58e5d6a5de2919a7422caf causes [JENKINS-69100](https://issues.jenkins.io/browse/JENKINS-69110) and undoes a lot of work done on this PR.

Bananeweizen commented:
Just to want to mention that this change has one big drawback: You now have to scroll both to the top and the bottom of the page to actually do updates. Previously both UI elements for selection and triggering were next to each other. Maybe that can be improved by adding the checkbox a second time at the bottom of the table?

uhafner commented:
Another option would be to use the table content as scrollable area and let the controls (and the Jenkins footer) always be visible at the screen bottom. That would make sense for a lot of other pages as well.

daniel-beck commented:
@Bananeweizen Please file this as an issue in Jira, or file a pull request implementing your suggestion.
